_overviews/scala3-book/tasty-for-end-users.md

```diff
@@ -0,0 +1,150 @@
+
+```scala
+val xs: List[Int] = List(1, 2, 3)
+```
+
+that code is compiled to a _.class_ file that needs to be compatible with the JVM. As a result of that compatibility requirement, the code inside that class file --- which you can see with a `javap` command --- ends up looking like this:
+
+```java
+public scala.collection.immutable.List<java.lang.Object> xs();
+```
+
+That `javap` command output shows a Java representation of what’s contained in the class file. Notice in this output that `xs` _is not_ defined as a `List[Int]` any more; it’s essentially represented as a `List[java.lang.Object]`. For your Scala code to work with the JVM, the `Int` type has been erased.
+
+Later, when you access an element of your `List[Int]` in your Scala code, like this:
+
+```scala
+val x = xs(0)
+```
+
+the resulting class file will have a casting operation for this line of code, which you can think of being like this:
+
+```
+int x = (Int) xs.get(0)               // Java-ish
+val x = xs.get(0).asInstanceOf[Int]   // more Scala-like
+```
+
+Again, this is done for compatibility, so your Scala code can run on the JVM.
+
+And this discussion only covers the topic of type erasure. There are similar issues for every other Scala construct that the JVM isn’t aware of, including constructs like unions, intersections, traits with parameters, and many more Scala 3 features.
+
+
+### Key points
+
+So that’s the first takeaway from this section: The types you specify in your Scala code aren’t represented completely accurately in _.class_ files.
+
+A second key point is to understand that there are differences between the information that’s available at _compile time_ and _run time_:
+
+- At **compile time**, as `scalac` reads and analyzes your code, it knows that `xs` is a `List[Int]`
+- When the compiler writes your code to a class file, it writes that `xs` is a `List[Object]`, and it adds casting information everywhere else `xs` is accessed
+- Then at **run time** --- with your code running inside the JVM --- the JVM doesn’t know that your list is a `List[Int]`
+
+With Scala 3 and Tasty, here’s another important note about compile time:
+
+- When you write Scala 3 code that uses other Scala 3 libraries, `scalac` doesn’t have to read their _.class_ files any more; it can read their _.tasty_ files, which, as mentioned, are an _exact_ representation of your code. This helps in many ways.
+
+{% comment %}
+NOTE: In regards to the previous paragraphs, I’ve read that there’s also metadata in the class files, but I don’t know those details.
+{% endcomment %}
+
+
+
+## Tasty benefits
+
+As you can imagine, having a complete representation of your code in a platform-neutral form has many benefits:
+
+{% comment %}
+TODO: These lines mostly come from https://www.scala-lang.org/blog/2018/04/30/in-a-nutshell.html, and I don’t know exactly what the first bullet point means:
+{% endcomment %}
+
+- The compiler uses it to support separate compilation.
```

Comment:
Regarding your comment: AFAIK the idea is the following: storing typed abstract syntax trees as intermediate results allows you to reuse them, instead of having to read and typecheck scala files from scratch. This is particularly important for features like inline and macros, that require access to the original typed tree, which is  not present in class files.